lightning/src/ln/channelmanager.rs

```diff
@@ -433,17 +435,48 @@ pub struct ChannelManager<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref,
 	/// new channel.
 	per_peer_state: RwLock<HashMap<PublicKey, Mutex<PeerState>>>,
 
+	persistence_notifier: PersistenceNotifier,
+
 	pending_events: Mutex<Vec<events::Event>>,
 	/// Used when we have to take a BIG lock to make sure everything is self-consistent.
 	/// Essentially just when we're serializing ourselves out.
 	/// Taken first everywhere where we are making changes before any other locks.
+	/// When acquiring this lock in read mode, rather than acquiring it directly, call
+	/// `ScopeGuardLockerNotifier::new(..)` and pass the lock to it, to ensure its desired
+	/// callback gets executed when the lock is released.
 	total_consistency_lock: RwLock<()>,
 
 	keys_manager: K,
 
 	logger: L,
 }
 
+/// Whenever we release the `ChannelManager`'s `total_consistency_lock`, from read mode, it is
+/// desirable to notify any listeners on `wait_timeout`/`wait` that new updates are available for
+/// persistence. Therefore, this struct is responsible for locking the total consistency lock and,
+/// upon going out of scope, sending the aforementioned notification (since the lock being released
+/// indicates that the updates are ready for persistence).
+struct ScopeGuardLockerNotifier<'a> {
+	persistence_notifier: &'a PersistenceNotifier,
+	// We hold onto this result so the lock doesn't get released immediately.
+	_lock_result: RwLockReadGuard<'a, ()>,
+}
+
+impl<'a> ScopeGuardLockerNotifier<'a> {
+	fn new(lock: &'a RwLock<()>, persistence_notifier: &'a PersistenceNotifier) -> Self {
+		Self {
+			persistence_notifier,
+			_lock_result: lock.read().unwrap(),
+		}
+	}
+}
+
+impl<'a> Drop for ScopeGuardLockerNotifier<'a> {
+	fn drop(&mut self) {
+		self.persistence_notifier.notify();
```

Comment:
Ah, right, yes, we probably should - the writing thread takes a write lock right afterwards so it would be nice (if its not a huge difference - the condvar lock is a much more immediate concern and the race here is probably much less likely to trigger).

Reply:
I'm getting `cannot move out of `self.read_guard` which is behind a mutable reference` from trying to release the lock early.

Reply:
I ran across the same problem with `JoinResult`. The workaround is to make member variable an `Option` and drop the result of `take()` instead.